syscall/syscall.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2022 The Ebiten Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build darwin
+// +build darwin
+
+package syscall
+
+import "unsafe"
+
+func SyscallX(fn, a1, a2, a3 uintptr) (r1, r2, err uintptr) {
+	return syscall_syscallX(fn, a1, a2, a3)
+}
+
+func Syscall6X(fn, a1, a2, a3, a4, a5, a6 uintptr) (r1, r2, err uintptr) {
+	return syscall_syscall6X(fn, a1, a2, a3, a4, a5, a6)
+}
+
+var syscall9XABI0 uintptr
+
+func syscall9X() // implemented in assembly
+
+func Syscall9X(fn, a1, a2, a3, a4, a5, a6, a7, a8, a9 uintptr) (r1, r2, err uintptr) {
+	args := struct{ fn, a1, a2, a3, a4, a5, a6, a7, a8, a9, r1, r2, err uintptr }{
+		fn, a1, a2, a3, a4, a5, a6, a7, a8, a9, r1, r2, err}
+	runtime_entersyscall()
+	runtime_libcCall(unsafe.Pointer(syscall9XABI0), unsafe.Pointer(&args))
+	runtime_exitsyscall()
+	return args.r1, args.r2, args.err
+}
+
+var syscallXFABI0 uintptr
+
+func syscallXF() // implemented in assembly
+
+func SyscallXF(fn, a1, a2, a3 uintptr, f1, f2, f3 float64) (r1, r2, err uintptr) {
```

Comment:
When is this necessary? Is there a counterpart in the runtime?

Reply:
`func Uniform1f(location int32, v0 float32)` uses it because it takes a float argument. It's the only use of it so it'd be nice to remove it. None of the tests call this function so I can't confirm if it's okay to just use SyscallX. What I know from the calling convention it is not.

 In C, the first float is expected to be in either F0 (on arm64) and X0 (on amd64) while integer arguments are expected to R0 and DI respectively. There is no counterpart in the go runtime. 

Perhaps a solution would be to put all the integers in the float register as well? This would mean not using the runtime's version of these functions though

Reply:
On Window's Syscall, float values can be passed with `math.Float32bits`. I was wondering if we really have to use a specific register, or Windows is special

Reply:
On Windows they put the arguments into the floating point registers like I suggested. [Line 46](https://go-review.googlesource.com/c/go/+/336550/15/src/runtime/sys_windows_amd64.s#46).

Reply:
Out of curiosity, how can the runtime distinguish the float and non-float arguments...? IIUC SyscallN only takes uintptr values.

Reply:
It doesn't distinguish between them. It assumes they are all floats and puts them in floating-point registers. This works since it is undefined what is in a register if it's not a parameter to the function.

Reply:
Aha, I see. In order to simplify the API, would it be nice to do the same way? Or, do you think we should have `SyscallXF` for performance or other issues? I prefer simpler APIs but I am not sure how much the drawbacks are.

Reply:
I thought a stack is used for stdcall rather than registers. Do we have to use registers here?

Reply:
I went ahead and implemented it all under SyscallN. The only draw back is a few extra instructions to move the parameters into the floating-point registers. It's definitely worth having the simpler API.

Stack is still used on x64 after 6 integer parameters.